The original file is provided by mpd, and
can be found in the Buildroot package directory.

Instead of creating an overlay file, having to check this manually whenever
changes are made within Buildroot, patch the master file with additional code

Upstream: n/a (Buildroot specific)

Signed-off-by: Andreas Ziegler <15275159+aeolio@users.noreply.github.com>
---
--- a/etc/init.d/S95mpd	2026-06-18 06:48:19.175438917 +0200
+++ b/etc/init.d/S95mpd	2026-08-31 07:12:10.000000000 +0200
@@ -8,13 +8,91 @@
 
 # Sanity checks
 [ -f /etc/$DAEMON.conf ] || exit 0
+[ -f /usr/bin/$DAEMON ] || exit 0
+
+HOTPLUG_DIR="/media"
+STATE_DIR="/var/lib/$DAEMON"
+
+IO_URING_MEMLIMIT=65536
+
+### initialize system variables
+# shellcheck source=/dev/null # do not follow included files
+[ -f /etc/nabla.conf ] && . /etc/nabla.conf
+
+### check if $DAEMON is configured to run
+[ "$NABLA_START_MUSICPLAYER" != "Y" ] && exit 0
+
+### functions that are used in more than one init script
+# shellcheck source=/dev/null # do not follow included files
+[ -x /etc/init.d/init_functions ] && . /etc/init.d/init_functions
+
+_config_value() {
+	awk '/'"$1"'/ {gsub("\"", ""); print $2}' /etc/mpd.conf
+}
+
+# directories
+CACHE_DIR="$(_config_value cache_directory)"
+PLIST_DIR="$(_config_value playlist_directory)"
+# files
+PIDFILE="$(_config_value pid_file)"
+PID_DIR="$(dirname "$PIDFILE")"
+LOGFILE="$(_config_value log_file)"
+LOG_DIR="$(dirname "$LOGFILE")"
+
+# unmute all alsa devices found on machine
+unmute() {
+	amixer="/usr/bin/amixer"
+	procfs="/proc/asound"
+
+	if [ -x "$amixer" ]; then
+		# loop through all sound devices
+		for card in "$procfs"/card[0-9]; do
+			# ignore non-USB devices
+			if [ ! -f "$card/usbid" ]; then continue; fi
+			# extract device id
+			id=${card##*/card}
+			# unmute all simple controls on this device
+			amixer -c "$id" scontrols | \
+				sed -e 's/^Simple mixer control//' | \
+				while read -r line; do 
+					amixer -c "$id" sset "$line" 100% unmute > /dev/null 2>&1
+				done
+		done
+	fi
+}
+
+# create directories and fix permissions
+# shellcheck disable=SC2174 # permissions changed only on last directoriy
+initialize() {
+	_owner="mpd:audio"
+	[ -n "$CACHE_DIR" ] && mkdir -m 755 -p "$CACHE_DIR" && chown "$_owner" "$CACHE_DIR"
+	[ -n "$PLIST_DIR" ] && mkdir -m 755 -p "$PLIST_DIR" && chown "$_owner" "$PLIST_DIR"
+	[ -n "$PID_DIR" ] && mkdir -m 755 -p "$PID_DIR" && chown "$_owner" "$PID_DIR"
+	[ -n "$LOG_DIR" ] && mkdir -m 755 -p "$LOG_DIR" && chown "$_owner" "$LOG_DIR"
+		# create ignore list
+	echo "$NABLA_PARTITION" > "$HOTPLUG_DIR"/.mpdignore
+	# library file attributes may be broken by import
+	find "$STATE_DIR" -type f -exec chmod u=rw,g=rw,o=r {} \;
+}
 
 start() {
 	printf "Starting %s: " "$DAEMON"
+	initialize
+	unmute
+	# increase max locked memory limit for io_uring 
+	# increase real-time priority limit to enable real-time scheduling
+	# shellcheck disable=SC3045 # busybox sh provides ulimit builtin
+	ulimit -HS -l "$IO_URING_MEMLIMIT" -r $((HARDIRQ_PRIORITY-1))
 	start-stop-daemon --start --pidfile "$PIDFILE" \
 		--exec "/usr/bin/$DAEMON"
 	status=$?
 	if [ "$status" -eq 0 ]; then
+		# set processor affinity
+		cpu_id=$(_get_rt_cpuid)
+		pid=$(cat "$PIDFILE")
+		taskset -p "$(_get_cpumask "$cpu_id")" "$pid"
+		# thread priority is set by process, so this might be unnecessary
+		chrt "$(_chrt_policy)" -p "$SOFTIRQ_PRIORITY" "$pid"
 		echo "OK"
 	else
 		echo "FAIL"
